File: Blind 75/01 - Arrays and Hashing/JavaScript/238-product-of-array-except-self.js
const { performance } = require("perf_hooks");

class Solution {
  method(nums) {
    let [prefix, suffix] = [1, 1];
    let res = [];
    for (const [i, n] of nums.entries()) {
      res[i] = prefix;
      prefix *= n;
    }

    for (let i = nums.length - 1; i >= 0; i--) {
      res[i] *= suffix;
      suffix *= nums[i];
    }

    return res;
  }

  reference(nums) {
    const result = [];
    let prefix = 1;
    let postfix = 1;

    for (let i = 0; i < nums.length; i++) {
      result[i] = prefix;
      prefix *= nums[i];
    }
    for (let i = nums.length - 2; i >= 0; i--) {
      postfix *= nums[i + 1];
      result[i] *= postfix;
    }

    return result;
  }

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.method(input));
        else this.method(input);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.reference(input));
        else this.reference(input);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = [
  [1, 2, 3, 4],
  [-1, 1, 0, -3, 3],
];
test.quantify(testCases);

File: Blind 75/01 - Arrays and Hashing/238-product-of-array-except-self.py
'''
Given an integer array nums, return an array answer such that answer[i] is equal to the product of all the elements of nums except nums[i].

The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit integer.

You must write an algorithm that runs in O(n) time and without using the division operation.
'''
from time import time
from typing import List


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        prefix = 1
        suffix = 1

        # Forward loop
        res = {}
        for i, e in enumerate(nums):
            res[i] = prefix
            prefix *= e

        # Reverse loop
        for i, e in reversed(list(enumerate(nums))):
            res[i] *= suffix
            suffix *= e

        return res.values()

    def reference(self, nums: List[int]) -> List[int]:
        res = [1] * (len(nums))

        prefix = 1
        for i in range(len(nums)):
            res[i] = prefix
            prefix *= nums[i]

        postfix = 1
        for i in range(len(nums) - 1, -1, -1):
            res[i] *= postfix
            postfix *= nums[i]

        return res

    def quantify(self, test_cases, runs=50000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.productExceptSelf(case))
                else:
                    self.productExceptSelf(case)
        print(f'Runtime for our solution: {time() - sol_start}')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(case))
                else:
                    self.reference(case)
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [[1, 2, 3, 4], [-1, 1, 0, -3, 3]]
    test.quantify(test_cases)
